File: day13/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	sum := 0
	sum2 := 0

	var grid [][]byte

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			grid = append(grid, []byte(scanner.Text()))
			continue
		}

		s, s2, ok := findReflections(grid)
		if !ok {
			log.Fatalln("No mirror")
		}
		sum += s
		sum2 += s2
		grid = nil
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	s, s2, ok := findReflections(grid)
	if !ok {
		log.Fatalln("No mirror")
	}
	sum += s
	sum2 += s2

	fmt.Println("Part 1:", sum)
	fmt.Println("Part 2:", sum2)
}

func findReflections(grid [][]byte) (int, int, bool) {
	tgrid := transpose(grid)
	s, eh, ev, ok := findReflection(grid, tgrid)
	if !ok {
		return 0, 0, false
	}
	s2, ok := findSmudgeReflection(grid, tgrid, eh, ev)
	if !ok {
		return 0, 0, false
	}
	return s, s2, true
}

func findReflection(grid, transpose [][]byte) (int, int, int, bool) {
	if hm := findMirror(grid); hm > 0 {
		return hm * 100, hm, -1, true
	}
	if vm := findMirror(transpose); vm > 0 {
		return vm, -1, vm, true
	}
	return 0, -1, -1, false
}

func findSmudgeReflection(grid, transpose [][]byte, eh, ev int) (int, bool) {
	if hm := findSmudge(grid, eh); hm > 0 {
		return hm * 100, true
	}
	if vm := findSmudge(transpose, ev); vm > 0 {
		return vm, true
	}
	return 0, false
}

func findSmudge(grid [][]byte, except int) int {
	for i := 1; i < len(grid); i++ {
		if i == except {
			continue
		}
		if isAlmostMirroredAt(grid, i) {
			return i
		}
	}
	return -1
}

func isAlmostMirroredAt(grid [][]byte, r int) bool {
	height := len(grid)
	lim := min(height-r, r)
	count := 0
	for i := 0; i < lim; i++ {
		a := grid[r-i-1]
		b := grid[r+i]
		if string(a) != string(b) {
			if !isEditDistance1(a, b) {
				return false
			}
			count += 1
			if count > 1 {
				return false
			}
		}
	}
	return count == 1
}

func isEditDistance1(a, b []byte) bool {
	count := 0
	for i := range a {
		if a[i] != b[i] {
			count += 1
		}
		if count > 1 {
			return false
		}
	}
	return count == 1
}

func findMirror(grid [][]byte) int {
	for i := 1; i < len(grid); i++ {
		if isMirroredAt(grid, i) {
			return i
		}
	}
	return -1
}

func isMirroredAt(grid [][]byte, r int) bool {
	height := len(grid)
	lim := min(height-r, r)
	for i := 0; i < lim; i++ {
		if string(grid[r-i-1]) != string(grid[r+i]) {
			return false
		}
	}
	return true
}

func transpose(grid [][]byte) [][]byte {
	height := len(grid)
	width := len(grid[0])
	res := make([][]byte, width)
	for i := range res {
		res[i] = make([]byte, height)
		for j := range res[i] {
			res[i][j] = grid[j][i]
		}
	}
	return res
}

File: day13/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut sum = 0;
    let mut sum2 = 0;

    let mut grid = Vec::new();

    for line in reader.lines() {
        let line = line?;
        if line != "" {
            grid.push(line.bytes().collect::<Vec<_>>());
            continue;
        }
        let (s, s2) = find_reflections(&grid).ok_or("No mirror")?;
        sum += s;
        sum2 += s2;
        grid = Vec::new();
    }

    let (s, s2) = find_reflections(&grid).ok_or("No mirror")?;
    sum += s;
    sum2 += s2;

    println!("Part 1: {}\nPart 2: {}", sum, sum2);

    Ok(())
}

fn find_reflections(grid: &Vec<Vec<u8>>) -> Option<(usize, usize)> {
    let tgrid = transpose(grid);
    let (s, eh, ev) = match find_reflection(grid, &tgrid) {
        Some(v) => v,
        None => return None,
    };
    let s2 = match find_smudge_reflection(grid, &tgrid, eh, ev) {
        Some(v) => v,
        None => return None,
    };
    Some((s, s2))
}

fn find_reflection(
    grid: &Vec<Vec<u8>>,
    transpose: &Vec<Vec<u8>>,
) -> Option<(usize, Option<usize>, Option<usize>)> {
    if let Some(v) = find_mirror(grid) {
        return Some((v * 100, Some(v), None));
    }
    if let Some(v) = find_mirror(transpose) {
        return Some((v, None, Some(v)));
    }
    None
}

fn find_smudge_reflection(
    grid: &Vec<Vec<u8>>,
    transpose: &Vec<Vec<u8>>,
    eh: Option<usize>,
    ev: Option<usize>,
) -> Option<usize> {
    if let Some(v) = find_smudge(grid, eh) {
        return Some(v * 100);
    }
    if let Some(v) = find_smudge(transpose, ev) {
        return Some(v);
    }
    None
}

fn find_smudge(grid: &Vec<Vec<u8>>, except: Option<usize>) -> Option<usize> {
    for i in 1..grid.len() {
        if let Some(v) = except {
            if i == v {
                continue;
            }
        }
        if is_almost_mirrored_at(grid, i) {
            return Some(i);
        }
    }
    None
}

fn is_almost_mirrored_at(grid: &Vec<Vec<u8>>, r: usize) -> bool {
    let height = grid.len();
    let lim = r.min(height - r);
    let mut has_diff = false;
    for i in 0..lim {
        let a = &grid[r - i - 1];
        let b = &grid[r + i];
        if a != b {
            if !is_edit_dist_1(a, b) {
                return false;
            }
            if has_diff {
                return false;
            }
            has_diff = true
        }
    }
    has_diff
}

fn is_edit_dist_1(a: &Vec<u8>, b: &Vec<u8>) -> bool {
    let mut has_diff = false;
    for (&a, &b) in a.iter().zip(b.iter()) {
        if a != b {
            if has_diff {
                return false;
            }
            has_diff = true
        }
    }
    has_diff
}

fn find_mirror(grid: &Vec<Vec<u8>>) -> Option<usize> {
    for i in 1..grid.len() {
        if is_mirrored_at(grid, i) {
            return Some(i);
        }
    }
    None
}

fn is_mirrored_at(grid: &Vec<Vec<u8>>, r: usize) -> bool {
    let height = grid.len();
    let lim = r.min(height - r);
    for i in 0..lim {
        if grid[r - i - 1] != grid[r + i] {
            return false;
        }
    }
    true
}

fn transpose(grid: &Vec<Vec<u8>>) -> Vec<Vec<u8>> {
    let height = grid.len();
    let width = grid[0].len();
    let mut res = vec![vec![0; height]; width];
    for i in 0..width {
        for j in 0..height {
            res[i][j] = grid[j][i];
        }
    }
    res
}
